tools/canonicalize-proto-file/Main.hs

```diff
@@ -126,6 +126,8 @@ instance CanonicalRank DotProtoMessagePart
     DotProtoMessageReserved _fs -> Left (Right ())
       -- We use '()' here because 'Canonicalize [DotProtoMessagePart]'
       -- collapses all of the 'DotProtoMessageReserved's into just one.
+    DotProtoMessageOption _ -> Left (Right ())
```

Comment:
So I'm new to this section of the code, too, but I think I know what we need to do here.  Basically, this module is formatting a `.proto` file with everything in a specified order.  The way that it works is that it maps each node in the syntax tree to some type that has an `Ord` instance and then uses that `Ord` instance to sort them.

This part of the code would probably be clearer and easier to fix if you first make a small refactor to replace the weird nested `Either` type with a new separate datatype, like this:

```haskell
data PartRank
    = PartRankDefinition (Int, DotProtoIdentifier)
    | PartRankReserved
    | PartRankField (Maybe FieldNumber)

instance CanonicalRank DotProtoMessagePart where
    canonicalRank = \case
      DotProtoMessageField f -> PartRankField (canonicalRank f)
      DotProtoMessageOneOf _ fs -> PartRankField (canonicalRank f)
      DotProtoMessageDefinition d -> PartRankDefinition (canonicalRank d)
      DotProtoMessageReserved _ -> PartRankReserved 
```

… and then once you refactor it like that then it's easier to add in a new constructor for message options, like this:

```haskell
data PartRank
    = PartRankDefinition (Int, DotProtoIdentifier)
    | PartRankReserved
    | PartRankField (Maybe FieldNumber)
    | PartRankOption DotProtoOption

instance CanonicalRank DotProtoMessagePart where
    canonicalRank = \case
      DotProtoMessageField f -> PartRankField (canonicalRank f)
      DotProtoMessageOneOf _ fs -> PartRankField (canonicalRank f)
      DotProtoMessageDefinition d -> PartRankDefinition (canonicalRank d)
      DotProtoMessageReserved _ -> PartRankReserved 
      DotProtoMessageOption o -> PartRankOption o
```

I haven't type-checked that suggestion, so there might be a few mistakes or other things that would need to be fixed, but I think that's the general idea of how to go about this.

Reply:
@Gabriel439 thanks for advise, it worked perfectly! Fixed in https://github.com/awakesecurity/proto3-suite/pull/143/commits/699277b85b5b449120e87c489706aa2391ce6c91

I also changed order of `PartRank` constructors to provide more logical ordering of protobuf AST.